src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -0,0 +1,390 @@
+
+    def setup(self, parameters):
+        """
+        _setup_
+
+        Create DAO Factory and setup some DAO.
+        """
+        myThread = threading.currentThread()
+        daofactory = DAOFactory(package="WMComponent.RucioInjector.Database",
+                                logger=self.logger, dbinterface=myThread.dbi)
+
+        self.getUninjected = daofactory(classname="GetUninjectedFiles")
+        self.getMigrated = daofactory(classname="GetMigratedBlocks")
+
+        self.findDeletableBlocks = daofactory(classname="GetDeletableBlocks")
+        self.markBlocksDeleted = daofactory(classname="MarkBlocksDeleted")
+        self.getUnsubscribed = daofactory(classname="GetUnsubscribedDatasets")
+        self.markSubscribed = daofactory(classname="MarkDatasetSubscribed")
+
+        daofactory = DAOFactory(package="WMComponent.DBS3Buffer",
+                                logger=self.logger, dbinterface=myThread.dbi)
+        self.setStatus = daofactory(classname="DBSBufferFiles.SetPhEDExStatus")
+        self.setBlockClosed = daofactory(classname="SetBlockClosed")
+
+    @timeFunction
+    def algorithm(self, parameters):
+        """
+        _algorithm_
+
+        Poll the database for uninjected files and attempt to inject them into
+        PhEDEx.
+        """
+        if not self.enabled:
+            logging.info("RucioInjector component is disabled in the configuration, exiting.")
+            return
+
+        logging.info("Running Rucio injector poller algorithm...")
+
+        try:
+            # files that failed to get their status updated in dbsbuffer
+            self._updateLFNState(self.filesToRecover, recovery=True)
+
+            # get dbsbuffer_file.in_phedex = 0
+            uninjectedFiles = self.getUninjected.execute()
+
+            # create containers in rucio  (and update local cache)
+            containersAdded = self.insertContainers(uninjectedFiles)
+            if self.containersCache.isCacheExpired():
+                self.containersCache.setCache(containersAdded)
+            else:
+                self.containersCache.addItemToCache(containersAdded)
+
+            # create blocks. Only update the cache once a rule gets created...
+            blocksAdded = self.insertBlocks(uninjectedFiles)
+
+            # now create a Rucio rule for every single block (and update local cache)
+            blocksAdded = self.insertBlockRules(blocksAdded)
+            if self.blocksCache.isCacheExpired():
+                self.blocksCache.setCache(blocksAdded)
+            else:
+                self.blocksCache.addItemToCache(blocksAdded)
+
+            # finally, create file replicas
+            self.insertReplicas(uninjectedFiles)
+
+            # now close blocks already uploaded to DBS
+            self.closeBlocks()
+
+            if self.lastRulesExecTime + self.pollRules <= int(time.time()):
+                self.deleteBlocks()
+                self.insertContainerRules()
```

Comment:
If we are making block rules what is the container rule for?

Reply:
Container rule is, I assume the data placement for the whole CMS dataset to put it at a single site. But I do wonder how you can do this in a single agent since this is really a systemwide thing.

Reply:
Eric is correct. The container rule needs to be created whenever a workflow is configured to make a cms dataset level subscription. It happens at any point in the lifetime of a workflow (usually in the first 24h of the workflow, once it goes to running-open/running-closed status).

Good comment about multiple subscriptions btw. Either PhEDExInjector has the necessary logic to see if there is already a dataset subscription for that. Or we duplicate subscriptions (harmless)

Reply:
Does this only get set for a single RSE? If you make a container rule on multiple RSEs then it would duplicate the whole dataset..
But here actually we can simplify the logic a lot with rucio: the desire is to have two copies of the output data during production for resilience and this is trivial with rucio if we define a single container rule (with grouping = DATASET) for the whole workflow with an RSE expression matching "all active production sites" and a replicas count = 2. Each source block that is injected will count towards one of the two and rucio will manage placing another as appropriate.

Reply:
single RSE. IF that's how we want to run production in the Rucio era, then we need to make sure there is a GH with the feature requirements.

Reply:
Maybe I don't understand this part of the code, but I thought this is called once per block update? I would expect the container to be created once for the whole workflow, is it handled here somehow once so that another block processing at a different site does not create also a container rule at that site?

Reply:
`insertContainerRules` is called once every many cycles of the component. It also only operates on output datasets (rucio containers) that have been configured at workflow level. Once the agent has created the container level rule, we persist that information in the local database and it's no longer triggered again.

About multiple agents, that's a fair point. What happens if the same rule gets created multiple times (same rucio acct, same container, same rse expression)? I believe it will return duplicate rule.
AFAICS, we need to have a smarter logic in place. I put this issue in place:
https://github.com/dmwm/WMCore/issues/9621

feel free to discuss anything that I missed over there.